#pragma once
#include "module.hpp"

#include "protobuf_messages/messages.pb.h"

namespace tests
{

using EchoProtocol = meta::protocol <
                     Type1Message,
                     Type2Message,
                     Type3Message,
                     Type4Message,
                     Type5Message,
                     Type6Message,
                     Type7Message,
                     Type8Message,
                     Type9Message,
                     Type10Message,
                     Type11Message,
                     Type12Message,
                     Type13Message,
                     Type14Message,
                     Type15Message,
                     Type16Message,
                     Type17Message,
                     Type18Message,
                     Type19Message,
                     Type20Message,
                     Type21Message,
                     Type22Message,
                     Type23Message,
                     Type24Message,
                     Type25Message,
                     Type26Message,
                     Type27Message,
                     Type28Message,
                     Type29Message,
                     Type30Message,
                     Type31Message,
                     Type32Message,
                     Type33Message,
                     Type34Message,
                     Type35Message,
                     Type36Message,
                     Type37Message,
                     Type38Message,
                     Type39Message,
                     Type40Message,
                     Type41Message,
                     Type42Message,
                     Type43Message,
                     Type44Message,
                     Type45Message,
                     Type46Message,
                     Type47Message,
                     Type48Message,
                     Type49Message,
                     Type50Message,
                     Type51Message,
                     Type52Message,
                     Type53Message,
                     Type54Message,
                     Type55Message,
                     Type56Message,
                     Type57Message,
                     Type58Message,
                     Type59Message,
                     Type60Message,
                     Type61Message,
                     Type62Message,
                     Type63Message,
                     Type64Message,
                     Type65Message,
                     Type66Message,
                     Type67Message,
                     Type68Message,
                     Type69Message,
                     Type70Message,
                     Type71Message,
                     Type72Message,
                     Type73Message,
                     Type74Message,
                     Type75Message,
                     Type76Message,
                     Type77Message,
                     Type78Message,
                     Type79Message,
                     Type80Message,
                     Type81Message,
                     Type82Message,
                     Type83Message,
                     Type84Message,
                     Type85Message,
                     Type86Message,
                     Type87Message,
                     Type88Message,
                     Type89Message,
                     Type90Message,
                     Type91Message,
                     Type92Message,
                     Type93Message,
                     Type94Message,
                     Type95Message,
                     Type96Message,
                     Type97Message,
                     Type98Message,
                     Type99Message,
                     Type100Message,
                     Type101Message,
                     Type102Message,
                     Type103Message,
                     Type104Message,
                     Type105Message,
                     Type106Message,
                     Type107Message,
                     Type108Message,
                     Type109Message,
                     Type110Message,
                     Type111Message,
                     Type112Message,
                     Type113Message,
                     Type114Message,
                     Type115Message,
                     Type116Message,
                     Type117Message,
                     Type118Message,
                     Type119Message,
                     Type120Message,
                     Type121Message,
                     Type122Message,
                     Type123Message,
                     Type124Message,
                     Type125Message,
                     Type126Message,
                     Type127Message,
                     Type128Message,
                     Type129Message,
                     Type130Message,
                     Type131Message,
                     Type132Message,
                     Type133Message,
                     Type134Message,
                     Type135Message,
                     Type136Message,
                     Type137Message,
                     Type138Message,
                     Type139Message,
                     Type140Message,
                     Type141Message,
                     Type142Message,
                     Type143Message,
                     Type144Message,
                     Type145Message,
                     Type146Message,
                     Type147Message,
                     Type148Message,
                     Type149Message,
                     Type150Message,
                     Type151Message,
                     Type152Message,
                     Type153Message,
                     Type154Message,
                     Type155Message,
                     Type156Message,
                     Type157Message,
                     Type158Message,
                     Type159Message,
                     Type160Message,
                     Type161Message,
                     Type162Message,
                     Type163Message,
                     Type164Message,
                     Type165Message,
                     Type166Message,
                     Type167Message,
                     Type168Message,
                     Type169Message,
                     Type170Message,
                     Type171Message,
                     Type172Message,
                     Type173Message,
                     Type174Message,
                     Type175Message,
                     Type176Message,
                     Type177Message,
                     Type178Message,
                     Type179Message,
                     Type180Message,
                     Type181Message,
                     Type182Message,
                     Type183Message,
                     Type184Message,
                     Type185Message,
                     Type186Message,
                     Type187Message,
                     Type188Message,
                     Type189Message,
                     Type190Message,
                     Type191Message,
                     Type192Message,
                     Type193Message,
                     Type194Message,
                     Type195Message,
                     Type196Message,
                     Type197Message,
                     Type198Message,
                     Type199Message,
                     SimpleClientMessage
                     >;

struct EchoServer : public module_base<EchoServer, EchoProtocol>
{
    template <typename Message>
    void onMessage(ClientConnection& conn, Message& msg)
    {
        send_message(conn, msg);
    }
};

} // namespace tests
